phpstan.neon.dist

```diff
@@ -11,8 +11,9 @@ services:
 		tags:
 			- phpstan.broker.dynamicStaticMethodReturnTypeExtension
 parameters:
-	level: 3
+	level: 4
 	inferPrivatePropertyTypeFromConstructor: true
+	treatPhpDocTypesAsCertain: false
```

Comment:
@westonruter Need suggestion. Can we keep this or do we need to remove it?

After removing this parameter, the Below error are occurring. most of it is about updating param type in function doc. But I believe updating it might lead to a less accurate description of param type. 

```log
------ ------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   includes/sanitizers/class-amp-base-sanitizer.php                                                                                                 
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 
  297    Strict comparison using === between null and string will always evaluate to false.                                                               
  730    Elseif condition is always true.                                                                                                                 
         💡 Because the type is coming from a PHPDoc, you can turn off this check by setting treatPhpDocTypesAsCertain: false in your phpstan.neon.dist.   
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ---------------------------------------------------- 
  Line   includes/sanitizers/class-amp-iframe-sanitizer.php  
 ------ ---------------------------------------------------- 
  431    Result of && is always false.                       
 ------ ---------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   includes/sanitizers/class-amp-link-sanitizer.php                                                                                                 
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 
  150    Right side of && is always true.                                                                                                                 
         💡 Because the type is coming from a PHPDoc, you can turn off this check by setting treatPhpDocTypesAsCertain: false in your phpstan.neon.dist.   
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   includes/utils/class-amp-image-dimension-extractor.php                                                                                           
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 
  281    If condition is always false.                                                                                                                    
         💡 Because the type is coming from a PHPDoc, you can turn off this check by setting treatPhpDocTypesAsCertain: false in your phpstan.neon.dist.   
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   src/Admin/ReaderThemes.php                                                                                                                       
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 
  354    Result of && is always false.                                                                                                                    
         💡 Because the type is coming from a PHPDoc, you can turn off this check by setting treatPhpDocTypesAsCertain: false in your phpstan.neon.dist.   
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   src/BackgroundTask/MonitorCssTransientCaching.php                                                                                                
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 
  137    Instanceof between array and DateTimeInterface will always evaluate to false.                                                                    
         💡 Because the type is coming from a PHPDoc, you can turn off this check by setting treatPhpDocTypesAsCertain: false in your phpstan.neon.dist.   
  137    Result of && is always false.                                                                                                                    
         💡 Because the type is coming from a PHPDoc, you can turn off this check by setting treatPhpDocTypesAsCertain: false in your phpstan.neon.dist.   
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   src/BackgroundTask/RecurringBackgroundTask.php                                                                                                   
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 
  79     Result of && is always false.                                                                                                                    
         💡 Because the type is coming from a PHPDoc, you can turn off this check by setting treatPhpDocTypesAsCertain: false in your phpstan.neon.dist.   
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   src/Embed/HandlesGalleryEmbed.php (in context of class AMP_Gallery_Block_Sanitizer)                                                              
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 
  48     Right side of && is always true.                                                                                                                 
         💡 Because the type is coming from a PHPDoc, you can turn off this check by setting treatPhpDocTypesAsCertain: false in your phpstan.neon.dist.   
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   src/Embed/HandlesGalleryEmbed.php (in context of class AMP_Gallery_Embed_Handler)                                                                
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 
  48     Right side of && is always true.                                                                                                                 
         💡 Because the type is coming from a PHPDoc, you can turn off this check by setting treatPhpDocTypesAsCertain: false in your phpstan.neon.dist.   
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------- 
  Line   src/Support/SupportData.php                            
 ------ ------------------------------------------------------- 
  1037   Unreachable statement - code above always terminates.  
 ------ ------------------------------------------------------- 

                                                                                                                        
 [ERROR] Found 12 errors                                                                                                
                                                                                                                        
```

Reply:
```
  730    Elseif condition is always true.                                                                                                                 
         💡 Because the type is coming from a PHPDoc, you can turn off this check by setting treatPhpDocTypesAsCertain: false in your phpstan.neon.dist.   
```

This line also seems to be in error:

https://github.com/ampproject/amp-wp/blob/0db08f0b40fcfddfab6b42554bf4fa9079dccad9/includes/sanitizers/class-amp-base-sanitizer.php#L730

It could indeed be replaced with:

```php
} else { // if ( $is_inline_style )
```

Reply:
```
 ------ ---------------------------------------------------- 
  Line   includes/sanitizers/class-amp-iframe-sanitizer.php  
 ------ ---------------------------------------------------- 
  431    Result of && is always false.                       
 ------ ---------------------------------------------------- 
```

For this line:

https://github.com/ampproject/amp-wp/blob/0db08f0b40fcfddfab6b42554bf4fa9079dccad9/includes/sanitizers/class-amp-iframe-sanitizer.php#L430-L431

The issue is that the `$value` has phpdoc typeof `string`. This could be fixed by making `$value` of `mixed` type. Or if it is determined that the value is only ever `string` then the `! is_string( $value ) &&` part could be removed.

Reply:
I haven't checked the others, but from checking the first few it seems that these are indeed issues that warrant fixing. So:

```suggestion
```